### PR TITLE
Remove mypy exceptions part 4

### PR DESCRIPTION
**Pull request by albertkol**
### Change description
Rest of the files. With few exceptions left behind.

### How to test
Is mypy and pytest happy?

### CODE REVIEW

**Comment by albertkol** [comment]
Type checks is failing currently on this error. (happens twice)
image

Turned off pre-commit so I can push changes.

File: data_store/transformation/towns_fund/tf_transform_r3.py [hunk 1]
```diff
@@ -115,7 +114,13 @@ def extract_place_details(df_place: pd.DataFrame) -> pd.DataFrame:
     df_place.columns = [0, 1, 2]
 
     # fill in blank (merged) cells
-    field_names_first = [x := y if y is not np.nan else x for y in df_place[0]]  # noqa: F841,F821
+    field_names_first = []
+    x = ""  # Initialize x
```
Comment (samuelhwilliams) [inline data_store/transformation/towns_fund/tf_transform_r3.py:118]:
```python
# comment: this is a comment
```

Comment (albertkol) [reply data_store/transformation/towns_fund/tf_transform_r3.py:118]:
```
review: this is a review
```

Clever..

File: data_store/serialisation/data_serialiser.py [hunk 2]
```diff
@@ -178,13 +180,13 @@ class FundingCommentSchema(SQLAlchemySchema):
...
+    submission_id = auto_field(model=typing.cast(DeclarativeMeta, Submission), data_key="SubmissionID")
+    programme_id = auto_field(model=typing.cast(DeclarativeMeta, Programme), data_key="ProgrammeID")
+    project_id = auto_field(model=typing.cast(DeclarativeMeta, Project), data_key="ProjectID")
     comment = JSONBStringField(attribute="data_blob.comment", data_key="Comment")
-    project_name = auto_field(model=Project, data_key="ProjectName")
-    programme_name = auto_field(model=Programme, data_key="Place")
-    organisation_name = auto_field(model=Organisation, data_key="OrganisationName")
+    project_name = auto_field(model=typing.cast(DeclarativeMeta, Project), data_key="ProjectName")
+    programme_name = auto_field(model=typing.cast(DeclarativeMeta, Programme), data_key="Place")
+    organisation_name = auto_field(model=typing.cast(DeclarativeMeta, Organisation), data_key="OrganisationName")
```
[the remark this answers is not in the document]
Comment (albertkol) [reply data_store/serialisation/data_serialiser.py:189]: I will take one more look at this. I will leave the file exception if I cannot get the `auto_field` to work.

File: data_store/validation/towns_fund/fund_specific_validation/fs_validate_r4.py [hunk 3]
```diff
@@ -107,8 +105,8 @@ def validate_project_risks(workbook: dict[str, pd.DataFrame]) -> list["GenericFa
     projects_missing_risks = project_details_df[~project_with_risks_mask & ~completed_projects_mask]["Project ID"]
 
     if len(projects_missing_risks) > 0:
-        projects_missing_risks = list(projects_missing_risks)
-        projects_missing_risks.sort()
+        projects_missing_risks = pd.Series(list(projects_missing_risks))
```
[the remark this answers is not in the document]
Comment (samuelhwilliams) [reply data_store/validation/towns_fund/fund_specific_validation/fs_validate_r4.py:108]: Would rather see that three-line block just replaced with

```python

        project_numbers = [
            get_project_number_by_id(project_id, active_project_ids) for project_id in sorted(projects_missing_risks)
        ]
```

no reassignment, no changing sort behaviour?

File: data_store/validation/towns_fund/fund_specific_validation/fs_validate_r4.py [hunk 4]
```diff
@@ -634,14 +639,14 @@ def validate_project_progress(workbook: dict[str, pd.DataFrame]) -> list["Generi
...
-                    row_index=row.name,
+                    row_index=typing.cast(int, row.name),
                 )
             )
 
     return failures
 
 
-def validate_sign_off(workbook: dict[str, pd.DataFrame]) -> list[GenericFailure]:
+def validate_sign_off(workbook: dict[str, pd.DataFrame] | None) -> list[GenericFailure]:
```
[the remark this answers is not in the document]
Comment (samuelhwilliams) [reply data_store/validation/towns_fund/fund_specific_validation/fs_validate_r4.py:649]: why can it be none 😿

[the remark this answers is not in the document]
Comment (albertkol) [reply data_store/validation/towns_fund/fund_specific_validation/fs_validate_r4.py:649]: `validate_sign_off(original_workbook)` -> `original_workbook` can be None when passed to `validate`:
```
def validate(
    data_dict: dict[str, pd.DataFrame], original_workbook: dict[str, pd.DataFrame] | None = None
) -> list[GenericFailure]:
```

File: data_store/validation/towns_fund/fund_specific_validation/fs_validate_r4.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (albertkol) [reply data_store/validation/towns_fund/fund_specific_validation/fs_validate_r4.py:108]: Good suggestion.

File: data_store/serialisation/data_serialiser.py [hunk 5]
```diff
@@ -23,11 +23,13 @@
 
 """
 
+import typing
```
[the remark this answers is not in the document]
Comment (albertkol) [reply data_store/serialisation/data_serialiser.py:26]: I reverted the whole change.

`data_serialiser` file remains excluded for `arg-type` error code.

File: data_store/serialisation/data_serialiser.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (albertkol) [reply data_store/serialisation/data_serialiser.py:189]: Reverted.

File: data_store/transformation/towns_fund/tf_transform_r3.py [hunk 6]
```diff
@@ -343,7 +356,7 @@ def extract_project_progress(df_data: pd.DataFrame, project_lookup: dict, report
     """
     # if round 4 or 5, ingest two additional columns
     df_data = df_data.iloc[18:39, 2:15] if reporting_round in (4, 5) else df_data.iloc[18:39, 2:13]
-    df_data = df_data.rename(columns=df_data.iloc[0]).iloc[1:]
+    df_data = df_data.rename(columns=df_data.iloc[0].to_dict()).iloc[1:]
```
Comment (samuelhwilliams) [inline data_store/transformation/towns_fund/tf_transform_r3.py:359]: 👀

File: data_store/transformation/towns_fund/tf_transform_r3.py [hunk 7]
```diff
@@ -453,8 +474,14 @@ def extract_funding_questions(df_input: pd.DataFrame, programme_id: str) -> pd.D
...
     fund_questions_df["original_index"] = fund_questions_df.index
     fund_questions_df.reset_index(drop=True, inplace=True)
-    fund_questions_df.loc[len(fund_questions_df)] = non_pivot_row
-    fund_questions_df.index = fund_questions_df["original_index"]
+
+    # Create a DataFrame from non_pivot_row
+    new_row = pd.DataFrame([non_pivot_row], columns=fund_questions_df.columns)
+
+    # Use pd.concat to add the new row
+    fund_questions_df = pd.concat([fund_questions_df, new_row], ignore_index=True)
```
Comment (samuelhwilliams) [inline data_store/transformation/towns_fund/tf_transform_r3.py:482]: how sure are we about this behaving the same?

File: data_store/transformation/towns_fund/tf_transform_r3.py [hunk 8]
```diff
@@ -635,7 +670,7 @@ def extract_funding_data(df_input: pd.DataFrame, project_lookup: dict, reporting
         mask, ["Funding Source Name", "Funding Source Type"]
     ].values
 
-    df_funding.index = df_funding["original_index"]
+    df_funding.index = pd.Index(df_funding["original_index"])
```
Comment (samuelhwilliams) [inline data_store/transformation/towns_fund/tf_transform_r3.py:673]: What's the diff between assigning to `df.index` and `df.columns`? we seem to be setting pd.Index on both?

File: data_store/transformation/towns_fund/tf_transform_r3.py [hunk 9]
```diff
@@ -965,7 +1007,7 @@ def extract_outcomes(df_input: pd.DataFrame, project_lookup: dict, programme_id:
     outcomes_df["Actual/Forecast"] = outcomes_df["Reporting Period"].str.split("__").str[1]
     # split out start and end dates for financial years
     outcomes_df["Start_Date"] = (outcomes_df["Reporting Period"].str[15:19] + "-04-01").astype("datetime64[ns]")
-    outcomes_df["End_Date"] = outcomes_df["Start_Date"] + MonthEnd(12)
+    outcomes_df["End_Date"] = outcomes_df["Start_Date"].apply(lambda x: x + pd.offsets.MonthEnd(12))
```
[the remark this answers is not in the document]
Comment (samuelhwilliams) [reply data_store/transformation/towns_fund/tf_transform_r3.py:1010]: definitely _a_ way

File: data_store/validation/towns_fund/schema_validation/validate.py [hunk 10]
```diff
@@ -61,19 +61,16 @@ def remove_undefined_tables(data_dict: dict[str, pd.DataFrame], schema: dict) ->
...
     """
     Validate the given data against a provided schema by checking each table's
     columns, data types, unique values, composite keys, and foreign keys.
 
-    :param data_dict: A dictionary where keys are table names and values are pandas
-                     DataFrames.
-    :param schema: A dictionary containing the validation schema for each table of the
-                   data.
+    :param data_dict: A dictionary where keys are table names and values are pandas DataFrames.
+    :param schema: A dictionary containing the validation schema for each table of the data.
```
Comment (samuelhwilliams) [inline data_store/validation/towns_fund/schema_validation/validate.py:70]: i find these kinds of comments increasingly less useful as we add type hints

**Review by samuelhwilliams — APPROVED** [review]

File: data_store/transformation/towns_fund/tf_transform_r3.py (hunk 7 above)
Comment (albertkol) [reply data_store/transformation/towns_fund/tf_transform_r3.py:482]: Ok.. I blindly relied on tests to and google for this one, as I am no pandas expert. 
But I have proof that it adds the new row at the end of the DataFrame as expected.

Before `fund_questions_df`:
```
                                                         Question  ... Programme ID
original_index                                                     ...             
14              Beyond these three funding types, have you rec...  ...       TD-FAK
16              Please confirm whether the amount utilised rep...  ...       TD-FAK
16              Please confirm whether the amount utilised rep...  ...       TD-FAK
16              Please confirm whether the amount utilised rep...  ...       TD-FAK
17              Please describe when funding was utilised and,...  ...       TD-FAK
17              Please describe when funding was utilised and,...  ...       TD-FAK
17              Please describe when funding was utilised and,...  ...       TD-FAK
19              Please explain in detail how the funding has, ...  ...       TD-FAK
19              Please explain in detail how the funding has, ...  ...       TD-FAK
19              Please explain in detail how the funding has, ...  ...       TD-FAK
15              Please indicate how much of your allocation ha...  ...       TD-FAK
15              Please indicate how much of your allocation ha...  ...       TD-FAK
15              Please indicate how much of your allocation ha...  ...       TD-FAK
18              Please select the option that best describes h...  ...       TD-FAK
18              Please select the option that best describes h...  ...       TD-FAK
18              Please select the option that best describes h...  ...       TD-FAK

[16 rows x 5 columns]
```

After `fund_questions_df`:
```
                                                         Question  ... Programme ID
original_index                                                     ...             
14              Beyond these three funding types, have you rec...  ...       TD-FAK
16              Please confirm whether the amount utilised rep...  ...       TD-FAK
16              Please confirm whether the amount utilised rep...  ...       TD-FAK
16              Please confirm whether the amount utilised rep...  ...       TD-FAK
17              Please describe when funding was utilised and,...  ...       TD-FAK
17              Please describe when funding was utilised and,...  ...       TD-FAK
17              Please describe when funding was utilised and,...  ...       TD-FAK
19              Please explain in detail how the funding has, ...  ...       TD-FAK
19              Please explain in detail how the funding has, ...  ...       TD-FAK
19              Please explain in detail how the funding has, ...  ...       TD-FAK
15              Please indicate how much of your allocation ha...  ...       TD-FAK
15              Please indicate how much of your allocation ha...  ...       TD-FAK
15              Please indicate how much of your allocation ha...  ...       TD-FAK
18              Please select the option that best describes h...  ...       TD-FAK
18              Please select the option that best describes h...  ...       TD-FAK
18              Please select the option that best describes h...  ...       TD-FAK

[16 rows x 5 columns]
```
(no, I did not paste the same thing twice to prove a point)

Let me know if you want the verbose, line by line, logs.

File: data_store/transformation/towns_fund/tf_transform_r3.py (hunk 8 above)
Comment (samuelhwilliams) [reply data_store/transformation/towns_fund/tf_transform_r3.py:673]: thanks for explaining :)

File: data_store/transformation/towns_fund/tf_transform_r3.py (hunk 6 above)
Comment (albertkol) [reply data_store/transformation/towns_fund/tf_transform_r3.py:359]: `to_dict` made mypy happy. It's a small price to pay to make `mypy` happy.

It takes something that looks like this:
```
2     B1\nTD 5% CDEL Pre-Payment; RDEL Capacity Fund...
4     TD 5% CDEL Pre-Payment\n(Towns Fund FAQs p.46 ...
5                              TD RDEL Capacity Funding
8                                TD Accelerated Funding
12                                       Guidance Notes
Name: 13, dtype: object
```
Converts it to a dict:
```
{
    2: 'B1\nTD 5% CDEL Pre-Payment; RDEL Capacity Funding; and Accelerated Funding',
    4: 'TD 5% CDEL Pre-Payment\n(Towns Fund FAQs p.46 - 49)',
    5: 'TD RDEL Capacity Funding',
    8: 'TD Accelerated Funding',
    12: 'Guidance Notes'
}
```

The result is still the same. Converts the 1st row of a DataFrame to it's header.
Before (note that header is [2, 12...]):
```
                                                   2   ...                                                 12
13  B1\nTD 5% CDEL Pre-Payment; RDEL Capacity Fund...  ...                                     Guidance Notes
14  Beyond these three funding types, have you rec...  ...                                                NaN
15  Please indicate how much of your allocation ha...  ...                                                NaN
16  Please confirm whether the amount utilised rep...  ...                                                NaN
17  Please describe when funding was utilised and,...  ...                                                NaN
18  Please select the option that best describes h...  ...  Given your answer on the 5% CDEL Pre-Payment, ...
19  Please explain in detail how the funding has, ...  ...                                                NaN

[7 rows x 5 columns]
```
After:
```
   B1\nTD 5% CDEL Pre-Payment; RDEL Capacity Funding; and Accelerated Funding  ...                                     Guidance Notes
14  Beyond these three funding types, have you rec...                          ...                                                NaN
15  Please indicate how much of your allocation ha...                          ...                                                NaN
16  Please confirm whether the amount utilised rep...                          ...                                                NaN
17  Please describe when funding was utilised and,...                          ...                                                NaN
18  Please select the option that best describes h...                          ...  Given your answer on the 5% CDEL Pre-Payment, ...
19  Please explain in detail how the funding has, ...                          ...                                                NaN

[6 rows x 5 columns]
```

This is from line 461 where we do this. It's easier to see the change.

**Pull request merged**